### ISSUE TITLE
What means active_wait option in pfring_zc_run_balancer?

### ISSUE THREAD

**Issue by Skryabind** [ntop/PF_RING#78]
What difference between active and passive wait mode in balancer?
If I set active_wait to 0, then my program doesn't receive all traffic on the interface: all incoming traffic is about 3Gbps, but program receives only 2.2Gbps. In this case CPU load is small on all cores.
If I set active_wait to 1, then program receive all traffic, but I see high cpu load on one core (up to 100%).
Driver - ixgbe, CPU: 2x Intel(R) Xeon(R) CPU E5-2630 0 @ 2.30GHz, total 24 cores.

**Comment by Skryabind** [comment]
Some info.
When I run 
`/zbalance -i zc:eth6 -c 1 -g 1:2:3:4:5:6:7:8:9:10:11:12:13:14:15:16:17:18:19:20`
I get next stats:
`Actual Stats: 281'250.00 pps - 2.56 Gbps`
but in active wait mode:
`/zbalance -i zc:eth6 -c 1 -g 1:2:3:4:5:6:7:8:9:10:11:12:13:14:15:16:17:18:19:20 -a`
I get more traffic:
`Actual Stats: 437'701.91 pps - 3.08 Gbps`

Is it a bug?

**Comment by cardigliano** [comment]
Hi
in active wait mode the CPU is spinning checking for new packets, otherwise, in passive mode, it waits using poll/usleep before checking for new packets.
It is strange that at that rate you are loosing packets even using passive mode, there should be something related to your hw/system configuration, any chance to get access to the machine?

Alfredo

**Comment by Skryabind** [comment]
Hello, Alfredo! Thank you for you answer!
I founded why some packets were lost. This is because in Debian CPU is in powersafe mode by default. So, when there is no load in CPU, system decrease frequency, and packets begins lose. If you load cpu, system increase frequency, and all packets will be received by program. I tested it in zbalance. In zbalance with active mode I receive all traffic because there is a high load on one core, so CPU works in maximum frequency.
So, there are interesting chain:
Small traffic -> low CPU load -> activating power safe mode in CPU -> losing traffic.
I thing this might be interesting issue for you and for people who uses pf_ring library. This is not ordinary problem.
Solution of this is setting CPU scaling profile to 'performance'.

1 To check you current CPU frequency:
apt-get install cpufrequtils sysfsutils
utility cpufreq-info show current CPU freq.

2 To check current CPU scaling profile:
cat /sys/devices/system/cpu/cpu0/cpufreq/scaling_governor
If there is a 'powersafe' profile, then you probably will get a problem like in this issue.

3 To set 'performance' profile:
echo performance > /sys/devices/system/cpu/cpu0/cpufreq/scaling_governor
Do it for each processor.
To set profile constantly, edit /etc/sysfs.conf

There are some other profiles, but 'performance' was the best for me.
